tiledb/sm/c_api/tiledb_filestore.cc

```diff
@@ -31,238 +31,260 @@
+#include "tiledb/type/range/range.h"
 
 namespace tiledb::common::detail {
 
 // Forward declarations
 uint64_t compute_tile_extent_based_on_file_size(uint64_t file_size);
-std::pair<Status, optional<std::string>> libmagic_get_mime(
-    void* data, uint64_t size);
-std::pair<Status, optional<std::string>> libmagic_get_mime_encoding(
-    void* data, uint64_t size);
+std::string libmagic_get_mime(void* data, uint64_t size);
+std::string libmagic_get_mime_encoding(void* data, uint64_t size);
 bool libmagic_file_is_compressed(void* data, uint64_t size);
-Status read_file_header(
-    const VFS& vfs, const char* uri, std::vector<char>& header);
+void read_file_header(
+    tiledb::sm::VFS& vfs, const char* uri, std::vector<char>& header);
 std::pair<std::string, std::string> strip_file_extension(const char* file_uri);
-std::pair<Status, optional<uint64_t>> get_buffer_size_from_config(
-    const Context& context, uint64_t tile_extent);
+uint64_t get_buffer_size_from_config(
+    const tiledb::sm::Config& config, uint64_t tile_extent);
 
 int32_t tiledb_filestore_schema_create(
     tiledb_ctx_t* ctx, const char* uri, tiledb_array_schema_t** array_schema) {
-  if (sanity_check(ctx) == TILEDB_ERR) {
-    *array_schema = nullptr;
-    return TILEDB_ERR;
-  }
-
-  Context context(ctx, false);
+  tiledb::sm::Context& context = ctx->context();
   uint64_t tile_extent = tiledb::sm::constants::filestore_default_tile_extent;
 
   bool is_compressed_libmagic = true;
   if (uri) {
     // The user provided a uri, let's examine the file and get some insights
     // Get the file size, calculate a reasonable tile extent
-    VFS vfs(context);
-    uint64_t file_size = vfs.file_size(std::string(uri));
+    tiledb::sm::VFS vfs(
+        &context.resources().stats(),
+        context.compute_tp(),
+        context.io_tp(),
+        context.storage_manager()->config());
+    uint64_t file_size;
+    throw_if_not_ok(vfs.file_size(tiledb::sm::URI(uri), &file_size));
     if (file_size) {
       tile_extent = compute_tile_extent_based_on_file_size(file_size);
     }
 
     // Detect if the file is compressed or not
     uint64_t size = std::min(file_size, static_cast<uint64_t>(1024));
     std::vector<char> header(size);
-    auto st = read_file_header(vfs, uri, header);
+
     // Don't fail if compression cannot be detected, log a message
     // and default to uncompressed array
-    if (st.ok()) {
+    try {
+      read_file_header(vfs, uri, header);
       is_compressed_libmagic = libmagic_file_is_compressed(header.data(), size);
-    } else {
-      LOG_STATUS_NO_RETURN_VALUE(
-          Status_Error("Compression couldn't be detected - " + st.message()));
+    } catch (const std::exception& e) {
+      LOG_STATUS_NO_RETURN_VALUE(Status_Error(
+          std::string("Compression couldn't be detected - ") + e.what()));
     }
   }
 
   *array_schema = new (std::nothrow) tiledb_array_schema_t;
   if (*array_schema == nullptr) {
```

Comment:
This means this check and throw doesn't make sense either, we should just let `operator new` throw